### PR TITLE
switch-next: Build with CMake

### PR DESCRIPTION
**Pull request by glebm**
I haven't tested the binary

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -69,7 +69,13 @@ jobs:
...
       - checkout
-      - run: make -j$(nproc)
+      - run: echo deb http://deb.debian.org/debian stretch-backports main > /etc/apt/sources.list.d/debian-backports.list
+      - run: apt-get update && apt-get install -y -t stretch-backports cmake
+      # Install cmake files (https://github.com/devkitPro/docker/issues/3)
+      - run: dkp-pacman -S --needed --noconfirm --quiet devkitpro-pkgbuild-helpers
+      - run: mkdir build
+      - run: cd build && cmake .. -DBINARY_RELEASE=ON -DNONET=ON -DCMAKE_TOOLCHAIN_FILE=../CMake/switch/devkita64-libnx.cmake
+      - run: cd build && make -j$(nproc)
       - store_artifacts: {path: ./release/devilutionx.nro, destination: devilutionx.nro}
```
Comment (AJenbo) [inline .circleci/config.yml]: No artifact files found at /root/repo/release/devilutionx.nro

Comment (glebm) [reply .circleci/config.yml]: Ah, right, they're placed in build now

Comment (glebm) [reply .circleci/config.yml]: Should be fixed now

**Comment by rsn8887** [comment]
The binary works, but the embedded title, author name, and version seem to be wrong: Title is Devilution Team (used to be "diablo-nx"), Author is "Unspecified Author" (should be "Devilution Team", and version is "0.5.0" (no clue what it should be).

The executable name is "DevilutionX.nro" instead of "diablo-nx.nro", but that might be intended?

File: CMakeLists.txt [hunk 2]
```diff
@@ -358,3 +386,12 @@ if(APPLE)
 
   include(CPack)
 endif()
+
+if(SWITCH)
+  set(APP_TITLE "DevilutionX")
+  set(APP_TITLE "Devilution Team")
```
Comment (rsn8887) [inline CMakeLists.txt]: This should be APP_AUTHOR

**Comment by rsn8887** [comment]
The error display requester also works, it shows a red "Missing date file" error window if .MPG file not found. I suppose it should be "data file", not "date file", maybe a typo. Also, on Switch, I cannot press OK in that requeter, it seems maybe the SDL_controller button input is not hooked up to cause OK in these new error requesters?

File: CMakeLists.txt (hunk 2 above)
Comment (glebm) [reply CMakeLists.txt]: Done, thanks :)

**Comment by glebm** [comment]
Thank you for testing!

> The error display requester also works, it shows a red "Date file error" window if the .MPG file is missing. I suppose it should be "data file", not "date file".

This typo is fixed on master but the `switch-next` branch hasn't been rebased yet.

> Also, on Switch, I cannot press OK in that requeter, it seems maybe the SDL_controller button input is not hooked up so I have no way to cause an OK in these new error requesters.

Thanks, we should fix this directly on master.

**Comment by glebm** [comment]
> Binary works, but APP_AUTHOR and APP_TITLE are wrong

This is fixed now. The version etc are consistent with the non-switch build (they weren't before).

**Comment by rsn8887** [comment]
I think PR should be ready to go AFAIK.

**Comment by rsn8887** [comment]
Although what happened to the line endings in `switch/switch_keyboard.cpp`? Is it intended? The diff shows the file completely changed, although there's only one tiny change. I think that's line endings then...

**Comment by glebm** [comment]
> Although what happened to the line endings in switch/switch_keyboard.cpp? Is it intended? The diff shows the file completely changed, although there's only one tiny change. I think that's line endings then...

Yeah, the line endings were wrong previously. It's not a big issue, as GitHub has a hide whitespace diff option.

**Comment by rsn8887** [comment]
Ok then this is ready to merge as far as I am concerned. Awesome work.

**Comment by AJenbo** [comment]
Thanks for testing, I'll rebase and replace the old PR

**Comment by AJenbo** [comment]
@glebm looks like this breaks after rebasing :(

**Comment by glebm** [comment]
> read-only segment has dynamic relocations

That's what I saw when LTO was disabled

**Comment by AJenbo** [comment]
Ok, that make a bit of sens since i switched it to DNIGHTLY_BUILD where that isn't being enabed. Just odd that Switch would need that to link successfully :/

**Comment by AJenbo** [comment]
But now if i switch to `-DBINARY_RELEASE=ON` it on i get:
`  LTO not supported by this compiler and/or CMake version`

**Comment by glebm** [comment]
Confirmed the same warning on CI, weird.

**Comment by glebm** [comment]
OK, I think I figured it out: The `project` definition must be above `check_ipo_supported()`, so that CMake knows which languages to check LTO for.

**Comment by glebm** [comment]
I've added this fix and others in https://github.com/diasurgical/devilutionX/pull/346

**Comment by AJenbo** [comment]
If LTO is required I think we need to make it so for the SWITCH case, someone compiling from source might easily not understand why the produced bin doesn't link. I also wonder what it is that the compiler is doing different compared to what was done in the make file? Maybe other flags are causing the issue.

**Comment by glebm** [comment]
> I also wonder what it is that the compiler is doing different compared to what was done in the make file?

Nothing different, it fails without LTO even with the makefile.

**Comment by glebm** [comment]
> If LTO is required I think we need to make it so for the SWITCH case

We could as a temporary workaround but ideally we should just fix the issue. No idea what's causing it though.

**Comment by AJenbo** [comment]
Yeah, I was thinking as a temp thing as we have no idea how to fix it :) Maybe the slow build time will annoy someone who knows what the issue is :D

**Comment by AJenbo** [comment]
> Nothing different, it fails without LTO even with the makefile.

But this issue isn't present in the parent branch, also build time is only 30sec vs 150sec, and I don't see any LTO flags in the Makefile, so it must have been something that was changed in this branch? The only change I see that would affect the make file build, in this PR, is the use of config.h and the change to StormPort.h.

**Comment by glebm** [comment]
Figured it out! The missing flag was `-fPIE`.

> build time is only 30sec vs 150sec

This is only due to LTO (we're linking way too much extra stuff for the switch, like OGG Vorbis etc). It's back to 30sec without it.

**Pull request merged**